mypy/checkexpr.py

```diff
@@ -2421,13 +2421,21 @@ def _get_value(self, index: Expression) -> Optional[int]:
                 operand = index.expr
                 if isinstance(operand, IntExpr):
                     return -1 * operand.value
+        typ = self.accept(index)
+        if isinstance(typ, LiteralType) and isinstance(typ.value, int):
+            return typ.value
         return None
 
     def visit_typeddict_index_expr(self, td_type: TypedDictType, index: Expression) -> Type:
-        if not isinstance(index, (StrExpr, UnicodeExpr)):
-            self.msg.typeddict_key_must_be_string_literal(td_type, index)
-            return AnyType(TypeOfAny.from_error)
-        item_name = index.value
+        if isinstance(index, (StrExpr, UnicodeExpr)):
+            item_name = index.value
+        else:
+            typ = self.accept(index)
+            if isinstance(typ, LiteralType) and isinstance(typ.value, str):
+                item_name = typ.value
+            else:
+                self.msg.typeddict_key_must_be_string_literal(td_type, index)
+                return AnyType(TypeOfAny.from_error)
```

Comment:
Maybe not really a question for this diff, but is there a reason for this to not live in  the typeddict plugin as a `__getitem__` hook?

There is maybe a more relevant question of: could this just use `try_getting_str_literal`, the answer to which would be more obvious if this lived in the plugin.

Reply:
I agree -- I think we should move both this and `tuple.__getitem__` out into a plugin, but it felt like it'd be out-of-scope for this PR.